### PR TITLE
Kubeadm: kustomize core

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR introduces a new feature that allows applying kustomize patches to static pod manifests generated by kubeadm. See [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/kubeadm/20190722-Advanced-configurations-with-kubeadm-(Kustomize).md)

**Which issue(s) this PR fixes**:
xref #https://github.com/kubernetes/kubeadm/issues/1379

**Special notes for your reviewer**:
WRT to the POC [https://github.com/kubernetes/kubernetes/pull/80580] this PR:
- contains only the core part of the kustomize implementation in kubeadm (UX changes will be implemented as a follow up PR starting from placeholders added in the second commit)
- address all the feedbacks collected in the POC
- adds unit test

Known limitations:
this first implementation as some known limitations:
- it does not work with customization stored in a git repository
- it does not support patchJson6902 patches

**Does this PR introduce a user-facing change?**:
```release-note
Implement a new feature that allows applying kustomize patches to static pod manifests generated by kubeadm. 
``` 

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/kubeadm/20190722-Advanced-configurations-with-kubeadm-(Kustomize).md
```

/sig cluster-lifecycle
/priority important-soon
/assign @rosti @yagonobre @neolit123

### CODE REVIEW

File: cmd/kubeadm/app/util/kustomize/kunstruct.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
+/*
```
Comment (ereslibre) [inline cmd/kubeadm/app/util/kustomize/kunstruct.go]: Minor: filename `kunstruct.go`? Could it be named `unstructured.go` instead? I think it's more clear.

File: cmd/kubeadm/app/util/kustomize/kustomize.go [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+		if err != nil {
+			return nil, err
+		}
+
+		name := fmt.Sprintf("resource-%d.json", i)
+		_ = memFS.WriteFile(filepath.Join(fakeDir, name), b)
+		fmt.Fprintf(&kustomization, " - %s\n", name)
+	}
+
+	kustomization.WriteString("patches:\n")
```
Comment (ereslibre) [inline cmd/kubeadm/app/util/kustomize/kustomize.go]: I think it's fine for now but would a template look more clear?

**Review by ereslibre — COMMENTED** [review]
Two minor comments, SGTM on a first pass.

File: cmd/kubeadm/app/phases/etcd/local.go [hunk 3]
```diff
@@ -25,7 +25,7 @@ import (
 	"github.com/pkg/errors"
 	"k8s.io/klog"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (yagonobre) [inline cmd/kubeadm/app/phases/etcd/local.go]: Why we are renaming this import?

**Review by yagonobre — COMMENTED** [review]
Thanks @fabriziopandini 
After address @ereslibre comments LGTM

File: cmd/kubeadm/app/phases/etcd/local.go (hunk 3 above)
Comment (neolit123) [reply cmd/kubeadm/app/phases/etcd/local.go]: goimports do does it automatically if i recall correctly.
there was a discussion about this in k/k but overall it's fine, as long as we don't sent huge PRs that modify all of these at once.

File: cmd/kubeadm/app/util/kustomize/kustomize.go (hunk 2 above)
Comment (fabriziopandini) [reply cmd/kubeadm/app/util/kustomize/kustomize.go]: I'm working on a follow-up PR for overcoming known limitations, and I will fix this in that PR

**Comment by fabriziopandini** [comment]
@yagonobre @ereslibre thanks for the feedback! Everything should be addressed/commented now

File: cmd/kubeadm/app/util/kustomize/kustomize.go [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+
+	// if there are no patches, for the target resources, exit
+	if len(patches) == 0 {
+		return res, nil
+	}
+
+	// create an in memory fs to use for the kustomization
+	memFS := fs.MakeFakeFS()
+
+	var kustomization bytes.Buffer
```
Comment (rosti) [inline cmd/kubeadm/app/util/kustomize/kustomize.go]: Ideally this should be `var kustomization kustomizetypes.Kustomization` (where `kustomizetypes "sigs.k8s.io/kustomize/pkg/types"`) and then simply marshal it to YAML.

File: cmd/kubeadm/app/util/kustomize/kustomize.go (hunk 2 above)
Comment (rosti) [reply cmd/kubeadm/app/util/kustomize/kustomize.go]: We already import some Kustomize packages, importing the types package, filling in a struct and marshaling it to YAML looks like the cleanest way to me.

**Review by rosti — COMMENTED** [review]
Thanks @fabriziopandini !
Overall, this looks good!

File: cmd/kubeadm/app/util/kustomize/kustomize.go (hunk 4 above)
Comment (fabriziopandini) [reply cmd/kubeadm/app/util/kustomize/kustomize.go]: Agreed, that's the approach I'm already following in the follow-up PR for overcoming known limitations

**Comment by ereslibre** [comment]
/lgtm
/hold

Adding hold for leaving time to address @rosti comments, feel free to unhold. I'm always unsure if a single lgtm will merge right away.

File: cmd/kubeadm/app/phases/controlplane/manifests_test.go [hunk 5]
```diff
@@ -138,6 +141,56 @@ func TestCreateStaticPodFilesAndWrappers(t *testing.T) {
 	}
 }
 
+func TestCreateStaticPodFilesKustomize(t *testing.T) {
+	// Create temp folder for the test case
+	tmpdir := testutil.SetupTempDir(t)
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/controlplane/manifests_test.go]: non-blocking: we should stop importing `k8s.io/kubernetes/cmd/kubeadm/test"` in `.../app`.

**Comment by neolit123** [comment]
ok, lets get this merge this and let @fabriziopandini send a follow up.
/hold cancel
/retest

File: cmd/kubeadm/app/phases/controlplane/manifests_test.go [hunk 6]
```diff
@@ -18,20 +18,23 @@ package controlplane
...
+
 	"k8s.io/apimachinery/pkg/util/sets"
 	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
 	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
 	"k8s.io/kubernetes/cmd/kubeadm/app/phases/certs"
-	authzmodes "k8s.io/kubernetes/pkg/kubeapiserver/authorizer/modes"
-
+	staticpodutil "k8s.io/kubernetes/cmd/kubeadm/app/util/staticpod"
 	testutil "k8s.io/kubernetes/cmd/kubeadm/test"
+	authzmodes "k8s.io/kubernetes/pkg/kubeapiserver/authorizer/modes"
```
Comment (liggitt) [inline cmd/kubeadm/app/phases/controlplane/manifests_test.go]: this reference was just removed in https://github.com/kubernetes/kubernetes/pull/80307, avoid reintroducing it

File: cmd/kubeadm/app/util/kustomize/kustomize.go [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+	us           UnstructuredSlice
+}
+
+var (
+	lock      = &sync.Mutex{}
+	instances = map[string]*Manager{}
+)
+
+// GetManager return the KustomizeManager singleton instance
+func GetManager(kustomizeDir string) (*Manager, error) {
```
Comment (liggitt) [inline cmd/kubeadm/app/util/kustomize/kustomize.go]: why is this being used as a singleton? is the returned manager threadsafe?

File: cmd/kubeadm/app/util/kustomize/kustomize.go [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+	}
+
+	km.us = us
+	return nil
+}
+
+// Kustomize apply a set of patches to a resource.
+// Portions of the kustomize logic in this function are taken from the kubernetes-sigs/kind project
+func (km *Manager) Kustomize(res []byte) ([]byte, error) {
+	// create a loader that mimics the behavior of kubectl kustomize
```
Comment (liggitt) [inline cmd/kubeadm/app/util/kustomize/kustomize.go]: what ensures this mimicked behavior doesn't drift from actual kustomize behavior?

Comment (liggitt) [reply cmd/kubeadm/app/util/kustomize/kustomize.go]: cc @monopole for insights on more consistent ways to make use of the kustomize capabilities

File: cmd/kubeadm/app/util/kustomize/kustomize.go [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+
+	var paths = []string{}
+	for _, file := range files {
+		if file.IsDir() {
+			continue
+		}
+		paths = append(paths, file.Name())
+	}
+
+	// Create a loader that mimics the behavior of kubectl kustomize, inclusing support for reading from
```
Comment (liggitt) [inline cmd/kubeadm/app/util/kustomize/kustomize.go]: cc @monopole
would appreciate a look at this approach

**Comment by liggitt** [comment]
Would like the CLI team's feedback on the invocation of kustomize to avoid drifting implementations.

More broadly, at first glance, this seems like it makes every change to the generated static manifests a potentially breaking change (even specifying a flag can break a user that was already adding that flag themselves). What happens if user-targeted customizations no longer match generated resources (name changes, etc). Are the customizations silently dropped or does that produce an error?

File: cmd/kubeadm/app/phases/controlplane/manifests_test.go (hunk 6 above)
Comment (fabriziopandini) [reply cmd/kubeadm/app/phases/controlplane/manifests_test.go]: Good catch. unfortunately https://github.com/kubernetes/kubernetes/pull/80307 removed this reference from `manifests.go` but not from `manifests_test.go`
I opened https://github.com/kubernetes/kubeadm/issues/1713 to track this cleanup

File: cmd/kubeadm/app/util/kustomize/kustomize.go (hunk 7 above)
Comment (fabriziopandini) [reply cmd/kubeadm/app/util/kustomize/kustomize.go]: kubeadm applies kustomizations in different points of the workflow, and using a singleton allows us to avoid to re-read patches every time.
The returned manager is thread-safe.

File: cmd/kubeadm/app/util/kustomize/kustomize.go (hunk 8 above)
Comment (fabriziopandini) [reply cmd/kubeadm/app/util/kustomize/kustomize.go]: Just a note waiting for @monopole review
Being kubeadm still in the main k/k repo, we are not allowed to import "k8s.io/cli-runtime/pkg/kustomize/k8sdeps", when kubeadm will move out from k/k (ETA 1.18), there will be room to further reduce risk of drift by using types defined in this package.

**Comment by fabriziopandini** [comment]
@liggitt thanks for comments. 
The risk of "change to the generated static manifests" was discussed in the KEP, and for the alpha release of this feature, it was decided to simply add AI to release notes for changes and to improve documentation for upgrades suggesting user to --dry-run before upgrades in case of patches.
The KEP introduces also the need for an improvement to kubeadm error management "ensuring that the node remains in a consistent state in case of errors", but this will be implemented in follow up PRs.

**Comment by timothysc** [comment]
> The risk of "change to the generated static manifests" was discussed in the KEP 

TL;DR It's a power user feature that has been asked for, and we plan to clause it as such.

**Comment by liggitt** [comment]
> > The risk of "change to the generated static manifests" was discussed in the KEP
> 
> TL;DR It's a power user feature that has been asked for, and we plan to clause it as such.

ack

**Comment by dims** [comment]
Let's reach out to @monopole for a review of how kustomize is used. no issues about vendoring it back in.

File: cmd/kubeadm/app/util/kustomize/kustomize.go (hunk 8 above)
Comment (monopole) [reply cmd/kubeadm/app/util/kustomize/kustomize.go]: Sorry about the delay, i was out for a bit.

There’s no formal notion of kustomize as a versioned library or service as yet, though it’s being worked on.  So the kustomize API is merely 1) the main program command line arguments and 2) the kustomization file and patches it references.

Ergo the recommended way to integrate kustomize into program foo is to have foo write a kustomization file and associated resources / patches to a file system, execute kustomize in a sub-process with the desired working directory, then capture stdout for whatever purpose (discarding it on error).  This is how kubebuilder uses it.  One can upgrade kustomize (the binary) without recompiling foo (modulo major version numbers).

This PR looks to be integrating kustomize into kubeadm in a fashion similar to how kustomize is integrated into kubectl, via the snapshot at k8s.io/cli-runtime/pkg/kustomize.  Doing this means kubeadm users don't have to install kustomize, but you’re tying yourself to 1) an old kustomize snapshot, and 2) _a code arrangement intended only for kubectl and not supported for others._

I.e. the only build we'll try not to break as we change cli-runtime/pkg/kustomize is the kubectl build.  So at some point the kubeadm build will break.  I don't think the fix will be hard (some new imports, new call styles, etc. without much change to overall concept), but it will presumably be annoying.

Comment (fabriziopandini) [reply cmd/kubeadm/app/util/kustomize/kustomize.go]: @monopole thanks for the valuable feedback!
My take away is that the current implementation has some trade-off that can be acceptable for the alpha release of this feature, but in the long term we should converge on one of the kubebuilder or the kubectl model, or be a more active part of the ongoing work around "kustomize as a versioned library". I'm open an issue in the k/kustomize repo to properly track this.

Unless there are no different opinions, I'm +1 to get this merged as it is because the above points are related only to implementation details, without any impact on the user experience

File: cmd/kubeadm/app/util/kustomize/kustomize.go (hunk 9 above)
Comment (neolit123) [inline cmd/kubeadm/app/util/kustomize/kustomize.go]: typo `inclusing`->`including`

File: cmd/kubeadm/app/util/kustomize/kustomize.go (hunk 8 above)
Comment (fabriziopandini) [reply cmd/kubeadm/app/util/kustomize/kustomize.go]: as per comment above https://github.com/kubernetes-sigs/kustomize/issues/1441

File: cmd/kubeadm/app/util/kustomize/kustomize.go (hunk 9 above)
Comment (fabriziopandini) [reply cmd/kubeadm/app/util/kustomize/kustomize.go]: fixed

**Comment by fabriziopandini** [comment]
@liggitt If I'm not wrong all the comment are now addressed or properly tracked

**Comment by liggitt** [comment]
/approve
for go.mod changes
